File: 프로그래머스/3/42627. 디스크 컨트롤러/디스크 컨트롤러.java
import java.util.*;


class Solution {
    public int solution(int[][] jobs) {
        int answer = 0, t = 0, i = 0;
        
        Arrays.sort(jobs, new Comparator<int[]>() {
            public int compare(int[] o1, int[] o2){
                if(o1[0] == o2[0])
                    return o1[1] - o2[1];
                return o1[0] - o2[0];
            }
        });
        
        PriorityQueue<Task> pq = new PriorityQueue<>();
        
        while(i < jobs.length || !pq.isEmpty()){
            if(pq.isEmpty() && jobs[i][0] > t) t = jobs[i][0];
            while(i < jobs.length && jobs[i][0] <= t) {
                pq.add(new Task(jobs[i][1], jobs[i][0]));
                i++;
            }
            if(!pq.isEmpty()){
                Task cur = pq.poll();
                t += cur.takeTime;
                answer += (t - cur.requestTime);
            }
        }
        
        return answer / jobs.length;
    }
}


class Task implements Comparable<Task> {
    public int takeTime;
    public int requestTime;
    
    public Task(int takeTime, int requestTime){
        this.takeTime = takeTime;
        this.requestTime = requestTime;
    }
    
    public int compareTo(Task t){
        if(this.takeTime != t.takeTime)
            return this.takeTime - t.takeTime;
        return this.requestTime - t.requestTime;
    }
}

File: 프로그래머스/3/42627. 디스크 컨트롤러/디스크 컨트롤러.cpp
#include <string>
#include <vector>
#include <queue>
#include <algorithm>
#include <iostream>

using namespace std;

int solution(vector<vector<int>> jobs) {
    int t = 0, answer = 0, i = 0;
    sort(jobs.begin(), jobs.end());
    priority_queue<pair<int,int>, vector<pair<int,int>>, greater<pair<int,int>>> pq;
    
    while(i < jobs.size() || !pq.empty()) {
        if(pq.empty() && jobs[i][0] > t) t = jobs[i][0];
        while(i < jobs.size() && jobs[i][0] <= t){
            pq.push({jobs[i][1], jobs[i][0]});
            i++;
        }
        if(!pq.empty()){
            auto x = pq.top(); pq.pop();
            t += x.first;
            answer += t - x.second;
        }
    }
    return answer / jobs.size();
}